### DISCUSSION TITLE
[Discussion] Moving away from lerna

### DISCUSSION THREAD

**Discussion by zakariaelas** [medusajs/medusa#1480] [RFC]
As discussed in this [issue](https://github.com/lerna/lerna/issues/2703) and made public by the author in the project's [readme](https://github.com/lerna/lerna#readme), lerna is no longer maintained. I think now might be a good time to look for alternatives to move away from lerna. 

### Lerna in Medusa
For Medusa, lerna has been mostly used for symlinking, task execution (i.e: building packages), and (pre)releasing packages. In the last few months, we started using yarn workspaces with lerna, which, in retrospect, can now be seen as an incremental migration. Currently, we also have an [open PR](https://github.com/medusajs/medusa/pull/1085) for using turborepo for task execution/scheduling.

### Current release/publishing process

Up until now, Medusa's release process has been a little bit manual and more or less looks like this: new features and bug fixes get merged to the `develop` branch, we fix a time during the week to branch out from `develop` to a new release branch, publish a canary version, do some testing, publish the branch to npm, then merge the release branch to `master` then onto `develop` (correct me if I'm wrong @olivermrbl). I think we can (and should) use this opportunity to improve our release process.

### Next steps

After moving to yarn workspaces and having an open PR for turborepo, I think what's left is to choose an alternative tool for publishing packages to npm. One alternative to lerna when it comes to publishing is [Changesets](https://github.com/changesets/changesets). It seems like a good candidate but I personally don't have any experience with it and it does seem to follow a different philosophy when it comes to publishing/releasing which we may or may not benefit from.

### Helpful resources

I thought I'd share a few resources here that might help mere mortals (like myself) who want to be involved in this discussion:
- [Overview on what a monorepo is, it's main features, and a comparison of the tools available out there](https://monorepo.tools/).
- [Overview on monorepo tooling and opinions on what's best in each class](https://dev.to/hipstersmoothie/javascript-monorepo-tooling-48b9). Author of the article is the main author and maintainer of [auto](https://github.com/intuit/auto).
- [Great introduction to Changesets and what the workflow looks like](https://dnlytras.com/blog/using-changesets/)

Thanks for reading and let me know what you guys think!

**Comment by adrien2p** [comment]
As discussed earlier, It would be nice to be able to enforce the commit spec (message, description, name, etc...) by using a tool like [commitlint](https://github.com/conventional-changelog/commitlint) as well as adding some pre-commit hook using something like [husky](https://github.com/typicode/husky) or something like that.

The purpose would be to get the change log and commits in general to follow a common pattern and keep it align with everything.

I would like to ear about what you think about that :)

**Reply by zakariaelas** [reply]
Totally, I think it's a great idea! Just like you briefly mentioned earlier, following a convention for commits would also lead to a cleaner changelog and allow us to use other tools like [standard-version](https://github.com/conventional-changelog/standard-version) (which I believe is what you used for medusa-extender). 

What do you think about using that as our publisher tool for medusa? Any feedback on that based on your experience?

**Reply by adrien2p** [reply]
yes this is exactly what I use and generate my change log automatically as well as the commit that comes with the new version. It includes everything needed as well as the description are very understandable since it is enforce by the lint. Let's see what come up with NX and see that later on :)

**Comment by jaruesink** [comment]
We use NX successfully in several of our projects, it makes handling large code bases easier. We’d love to come alongside your team and work with @adrien2p to help setup toolsets for the best development experience. These tools (like Adrien suggested too) would also help with automating change logs and keeping everything consistent as a large open source network. NX has some powerful generation tools also if you want to help people generate their own plugins or many other potential use cases for that.

**Reply by zakariaelas** [reply]
Very interesting @jaruesink! I've been mostly focusing on swapping out the "publishing" capabilities of lerna for some other tool (hence why I mentioned changesets), but it seems like NX is a really really solid alternative for monorepos and comes out with lots of features like code generation (just like you mentioned). Will definitely take a closer look right now and let you know! Thanks a ton for the suggestion, really appreciate the input 🙏

**Comment by drewdecarme** [comment]
I think the functional concept of using a monorepo should remain intact, but I would highly suggest looking into Yarn Berry... also known as Yarn 3.x. It's support for monorepos doesn't really differ from what exists with Lerna... it just doesn't have the robust publishing capabilities that Learna does.

I've been using it in production for several large scale enterprises and it scales wonderfully as well as is an immense upgrade from using Yarn 1.x. It's faster, more explicit and intentional an how dependencies are managed an added and isn't all too different from what we're used to using right now with Lerna.

From their documentation...
> Despite the appearances, the Yarn workspaces and Lerna don't compete. In fact, Lerna will use Yarn's workspaces if possible. In a sense, you can see Lerna as a high-level layer on top of the low-level Yarn implementation.

In addition, they have a concept of Zero Installs that really increases predictability, caching, development onboarding, and my personal favorite, CI installation time. I'd recommend really considering this if you want to 1) use your own publishing tool 2) maintain the sanity of your monorepo without a lot of re-learning or overhead

https://yarnpkg.com/features/workspaces - workspaces docs
https://yarnpkg.com/features/pnp - Plug-n-Play which also supports `node_modules`
